read-fonts/src/tables/layout/closure.rs

```diff
@@ -0,0 +1,209 @@
+            let tag = record.feature_tag();
+            if features.contains(tag) {
+                indices.insert(idx as u16);
+            }
+        }
+        self.feature_indices_filter = Some(indices);
+    }
+
+    pub(crate) fn script_visited(&mut self, s: &Script) -> bool {
+        if self.script_count > MAX_SCRIPTS {
+            return true;
+        }
+
+        self.script_count += 1;
+
+        let delta = (s.offset_data().as_bytes().as_ptr() as usize - self.table_head) as u32;
+        if self.visited_script.contains(delta) {
+            return true;
+        }
+        self.visited_script.insert(delta);
+        false
+    }
+
+    pub(crate) fn langsys_visited(&mut self, langsys: &LangSys) -> bool {
+        if self.langsys_count > MAX_LANGSYS {
+            return true;
+        }
+
+        self.langsys_count += 1;
+
+        let delta = (langsys.offset_data().as_bytes().as_ptr() as usize - self.table_head) as u32;
+        if self.visited_langsys.contains(delta) {
+            return true;
+        }
+        self.visited_langsys.insert(delta);
+        false
+    }
+
+    pub(crate) fn feature_indices_limit_exceeded(&mut self, count: u16) -> bool {
+        let (new_count, overflow) = self.feature_index_count.overflowing_add(count);
+        if overflow {
+            self.feature_index_count = MAX_FEATURE_INDICES;
+            return true;
+        }
+        self.feature_index_count = new_count;
+        new_count > MAX_FEATURE_INDICES
+    }
+}
+
+impl ScriptList<'_> {
+    /// Return a set of all feature indices underneath the specified scripts, languages and features
+    /// if no script is provided, all scripts will be queried
+    /// if no language is provided, all languages will be queried
+    /// if no feature is provided, all features will be queried
+    pub(crate) fn collect_features(
+        &self,
+        layout_table_head: usize,
+        feature_list: &FeatureList,
+        scripts: Option<&IntSet<Tag>>,
+        languages: Option<&IntSet<Tag>>,
+        features: Option<&IntSet<Tag>>,
+    ) -> Result<IntSet<u16>, ReadError> {
+        let mut out = IntSet::empty();
+        let mut c =
+            CollectFeaturesContext::new(features, layout_table_head, feature_list, &mut out);
+        let script_records = self.script_records();
+        let font_data = self.offset_data();
+        if scripts.is_none() {
+            // All scripts
+            for record in script_records {
+                let script = record.script(font_data)?;
+                script.collect_features(&mut c, languages)?;
+            }
+        } else {
+            let scripts = scripts.unwrap();
+            for tag in scripts.iter() {
+                let Some(idx) = self.index_for_tag(tag) else {
+                    continue;
+                };
+                let script = script_records[idx as usize].script(font_data)?;
+                script.collect_features(&mut c, languages)?;
+            }
+        }
```

Comment:
prefer matching to `is_none()` + `unwrap()`:

```suggestion
        match scripts {
            None => {
                for record in script_records {
                    let script = record.script(font_data)?;
                    script.collect_features(&mut c, languages)?;
                }
            }
            Some(scripts) => {
                for idx in scripts.iter().filter_map(|tag| self.index_for_tag(tag)) {
                    let script = script_records[idx as usize].script(font_data)?;
                    script.collect_features(&mut c, languages)?;
                }
            }
        }
```

Reply:
oh yeah, it's much better, thanks!